File: kotlin/lesson 4/QuickSort.kt
// https://leetcode.com/problems/sort-an-array/

class Solution {
    fun sortArray(nums: IntArray): IntArray {
        quickSort(nums)
        return nums
    }

    private fun quickSort(nums: IntArray) {
        shuffle(nums)
        sort(nums, 0, nums.size - 1)
    }

    private fun sort(nums: IntArray, lo: Int, hi: Int) {
        if (lo >= hi) return
        val pivot = partition(nums, lo, hi)
        sort(nums, lo, pivot - 1)
        sort(nums, pivot + 1, hi)
    }

    private fun partition(nums: IntArray, lo: Int, hi: Int): Int {
        val pivot = nums[lo]
        var i = lo
        var j = hi + 1
        while (true) {
            while (nums[++i] < pivot && i < hi) {}
            while (nums[--j] > pivot && j > lo) {}
            if (i >= j) break
            swap(nums, i, j)
        }
        swap(nums, lo, j)
        return j
    }

    private fun swap(nums: IntArray, a: Int, b: Int) {
        nums[a] = nums[b].also { nums[b] = nums[a] }
    }

    fun shuffle(nums: IntArray) {
        for (i in nums.indices) {
            val r = (0 until i + 1).random()
            swap(nums, i, r)
        }
    }
}

File: ruby/lesson 4/quick_sort.rb
# https://leetcode.com/problems/sort-an-array/

def sort_array(nums)
    quick_sort(nums)
    nums
end

def quick_sort(nums)
    shuffle(nums)
    sort(nums, 0, nums.length - 1)
end

def sort(nums, lo, hi)
    return if hi <= lo
    pivot = partition(nums, lo, hi)
    sort(nums, lo, pivot - 1)
    sort(nums, pivot + 1, hi)
end

def partition(nums, lo, hi)
    pivot = nums[lo]
    i = lo
    j = hi + 1
    loop do
        loop do
            i += 1
            break unless nums[i] < pivot && i < hi
        end
        loop do
            j -= 1
            break unless nums[j] > pivot && j > lo
        end
        break if i >= j
        swap(nums, i, j)
    end
    swap(nums, lo, j)
    j
end

def swap(nums, a, b)
    nums[a], nums[b] = nums[b], nums[a]
end

def shuffle(nums)
    (0...nums.length).each do |i|
        r = rand(0...i + 1)
        swap(nums, i, r)
    end
end
